crates/sats/src/slim_slice.rs

```diff
@@ -0,0 +1,1060 @@
+            });
+        };
+    };
+}
+
+/// Convert to `A` but panic if `x.len() > u32::MAX`.
+fn expect_fit<A, E, B: TryInto<A, Error = LenTooLong<E>>>(x: B) -> A {
+    x.try_into().map_err(|e| e.len).expect("length didn't fit in `u32`")
+}
+
+fn into_box<T, U: Into<Box<[T]>>>(x: U) -> Box<[T]> {
+    x.into()
+}
+
+// =============================================================================
+// Raw slice utility type
+// =============================================================================
+
+/// Implementation detail of the other types.
+/// Provides some convenience but users of the type are responsible
+/// for safety, invariants and variance.
+#[repr(packed)]
+struct SlimRawSlice<T> {
+    /// The length of the slice.
+    len: u32,
+    /// A valid pointer to the slice data.
+    ptr: NonNull<T>,
+}
+
+impl<T> SlimRawSlice<T> {
+    /// Split `self` into a raw pointer and the slice length.
+    fn split(self) -> (*mut T, usize) {
+        (self.ptr.as_ptr(), self.len as usize)
+    }
+
+    /// Dereferences this raw slice into a shared slice.
+    ///
+    /// SAFETY: `self.ptr` and `self.len` must satisfy `from_raw_parts`'s requirements.
+    #[allow(clippy::needless_lifetimes)]
+    unsafe fn deref<'a>(&'a self) -> &'a [T] {
+        let (ptr, len) = self.split();
+        // SAFETY: caller is responsible for these.
+        unsafe { slice::from_raw_parts(ptr, len) }
+    }
+
+    /// Dereferences this raw slice into a mutable slice.
+    ///
+    /// SAFETY: `self.ptr` and `self.len` must satisfy `from_raw_parts_mut`'s requirements.
+    #[allow(clippy::needless_lifetimes)]
+    unsafe fn deref_mut<'a>(&'a mut self) -> &'a mut [T] {
+        let (ptr, len) = self.split();
+        // SAFETY: caller is responsible for these.
+        unsafe { slice::from_raw_parts_mut(ptr, len) }
+    }
+
+    /// Creates the raw slice from a pointer to the data and a length.
+    ///
+    /// It is assumed that `len <= u32::MAX`.
+    /// The caller must ensure that `ptr != NULL`.
+    const fn from_len_ptr(len: usize, ptr: *mut T) -> Self {
+        // SAFETY: caller ensured that `!ptr.is_null()`.
+        let ptr = unsafe { NonNull::new_unchecked(ptr) };
+        let len = len as u32;
+        Self { ptr, len }
+    }
+}
+
+impl<T> Copy for SlimRawSlice<T> {}
+impl<T> Clone for SlimRawSlice<T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+// =============================================================================
+// Owned boxed slice
+// =============================================================================
+
+/// Provides a slimmer version of `Box<[T]>`
+/// using `u32` for its length instead of `usize`.
+#[repr(transparent)]
+pub struct SlimSliceBox<T> {
+    /// The representation of this boxed slice.
+    raw: SlimRawSlice<T>,
+    /// Marker to ensure covariance and dropck ownership.
+    owned: PhantomData<T>,
+}
+
+/// `SlimSlice<T>` is `Send` if `T` is `Send` because the data is owned.
```

Comment:
This (and many other places) reference SlimSlice rather than SlimSliceBoxed, and some other artifacts of a renaming I assume of other types in here.

I'm gonna let you find em all but I will try to find em as I go to.